File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
       unordered_set<ListNode*> sett;

       while(head!=NULL){
        if(sett.count(head)){
            return head;
        }

        sett.insert(head);
        head= head->next;
       }
       return NULL;
        
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        sett = set()
        while(head!=None):
            if(head in sett):
                return head

            sett.add(head)
            head = head.next
        return None
        
        